### PR TITLE
Fixed 404 caused by unexpected question mark

### PR DESCRIPTION
**Pull request by ghost**
Very simply, this pull request modifies certain files to escape symbols in filenames that may cause 404s unescaped.

### CODE REVIEW

**Comment by LefsFlarey** [comment]
LGTM

**Comment by tboerger** [comment]
Please reopen against master.

**Comment by bkcsoft** [comment]
Reopen

**Comment by strk** [comment]
Could you provide a testcase for the issue, or at least instructions to reproduce ?

**Comment by ghost** [comment]
@strk It's a simple case of creating a file with '?' in the filename.

File: modules/template/template.go [hunk 1]
```diff
@@ -97,7 +97,7 @@ func NewFuncMap() []template.FuncMap {
 		"MD5":               base.EncodeMD5,
 		"ActionContent2Commits": ActionContent2Commits,
 		"EscapePound": func(str string) string {
-			return strings.NewReplacer("%", "%25", "#", "%23", " ", "%20").Replace(str)
+			return strings.NewReplacer("%", "%25", "#", "%23", " ", "%20", "?", "%3F").Replace(str)
```
Comment (strk) [inline modules/template/template.go:100]: I don't understand here what the dark red color on percent sign means, do you ?

**Comment by bkcsoft** [comment]
Quick question, doesn't golang have a URL_escape? :)

**Comment by strk** [comment]
https://golang.org/pkg/net/url/#QueryEscape

File: modules/template/template.go (hunk 1 above)
Comment (ghost) [reply modules/template/template.go:100]: It's not important, it's just invalid highlighting.

**Comment by ghost** [comment]
Yet you can't use QueryEscape for filenames though.

**Comment by strk** [comment]
Could the code duplication be avoided by putting the replace code in a common function ?
The code is currently duplicated in NewFuncMap (template/template.go) and editFilePost (routers/repo/editor.go)

**Comment by strk** [comment]
Besides, I just tested reproducing and I go the 404 before _and_ after applying your patch

**Comment by ghost** [comment]
@strk can you tell me what you did to get the 404?

**Comment by strk** [comment]
I reloaded the URL I got the 404 from before the patch (url containing the `?` sign). Now I realize I should have probably reloaded the previous page ? Will try again later.

**Comment by strk** [comment]
Meanwhile please consider removing the code duplication, and maybe put the shared function under automatic testing

**Comment by strk** [comment]
Ok, reloading the _previous_ page gives the correct link. So LGTM (but code duplication would really be better to avoid)

**Comment by tboerger** [comment]
LGTM

File: modules/template/template.go (hunk 1 above)
Comment (lunny) [reply modules/template/template.go:100]: A better method is to define as a static replacer

``` Go
var replacerEscapePound = strings.NewReplacer("%", "%25", "#", "%23", " ", "%20", "?", "%3F")
```

And then replace the two palces of references.

**Comment by metalmatze** [comment]
LGTM

**Pull request merged**